### PR TITLE
[WIP] Add CycleReward

### PR DESCRIPTION
**Pull request by Abhinay1997**
Hi `RE-N-Y`,

Opening a PR to merge CycleReward into imscore. The following things are pending:

- [ ] Move checkpoints to RE-N-Y/CycleReward etc... for all three variants
- [ ] Compute MAE w.r.t original model. 
- [ ] Add docs in README
- [ ] Unsure if importing the BLIP model from imscore.imreward is fine. As this makes cyclereward the only model in the repo with non-standalone files.

The inconsistency comes from the preprocessing difference:
        1. The original codebase resizes first and later applies normalization to [0, 1]
        2. `imscore` first normalizes to [0,1] and then resizes the image.

Let me know if you're okay with merging this and changes if so.

Usage:

```bash
pip install git+https://github.com/Abhinay1997/imscore@cyclereward
```

```python
from imscore.cyclereward.model import CycleReward
available_models = ["NagaSaiAbhinay/CycleReward-Combo" "NagaSaiAbhinay/CycleReward-T2I", "NagaSaiAbhinay/CycleReward-I2T"]
model_id = available_models[0]
model = CycleReward.from_pretrained(model_id)

prompts = "a photo of a cat"
pixels = Image.open("cat.jpg")
pixels = np.array(pixels)
pixels = rearrange(torch.tensor(pixels), "h w c -> 1 c h w") / 255.0

# prompts and pixels should have the same batch dimension
# pixels should be in the range [0, 1]
# score == logits
score = model.score(pixels, prompts) # full differentiable reward)
```

### CODE REVIEW

**Comment by Abhinay1997** [comment]
Sure! Do note the source of discrepancy between the ported and original model. cyclereward does resize with interpolation before normalization whereas imscore does normalization followed by resize with interpolation. Let me know if you need any additional details or changes on the PR!

```python
from PIL import Image
import requests
from io import BytesIO
import torch
import numpy as np
from einops import rearrange

def load_image_from_url(url):
    response = requests.get(url)
    img = Image.open(BytesIO(response.content))
    return img

cat_image = load_image_from_url("https://github.com/RE-N-Y/imscore/blob/main/test/cat.png?raw=true")
pixels = np.array(cat_image)
pixels = rearrange(torch.tensor(pixels), "h w c -> 1 c h w") / 255.0
texts = ["A cat", "A white cat", "A black cat", "A white cat crouched and sitting on a wall"]

### CycleReward Combo
from imscore.cyclereward.model import CycleReward
model = CycleReward.from_pretrained('NagaSaiAbhinay/CycleReward-Combo')
model.score(pixels,texts)
#tensor([[-1.4697],
#        [-0.6329],
#        [-1.8794],
#        [-0.1646]], grad_fn=<AddmmBackward0>)

### CycleReward T2I trained reward model
from imscore.cyclereward.model import CycleReward
model = CycleReward.from_pretrained('NagaSaiAbhinay/CycleReward-T2I')
model.score(pixels,texts)
#tensor([[ 0.0683],
#        [ 0.1722],
#        [-0.4022],
#        [ 0.3133]], grad_fn=<AddmmBackward0>)

### CycleReward I2T trained reward model
from imscore.cyclereward.model import CycleReward
model = CycleReward.from_pretrained('NagaSaiAbhinay/CycleReward-I2T')
model.score(pixels,texts)
#tensor([[-1.6095],
#        [-0.9063],
#        [-2.0124],
#        [-0.6515]], grad_fn=<AddmmBackward0>)
```

[colab](https://colab.research.google.com/drive/1b0Br7viXUQs5mSv1IKGvABqk1DpaHPfb?usp=sharing)

**Comment by RE-N-Y** [comment]
Just ran it on my end, I think I am happy to merge this in if you can update the README (Usage / Preference Scorers section). I think it's fine to keep the checkpoints under your HF account. For the discrepancy part, I will make a separate issue for that.

**Comment by Abhinay1997** [comment]
Glad to hear that! I have updated the README, hopefully I haven't missed anything. Acknowledged on the checkpoint and the discrepancy!

**Comment by RE-N-Y** [comment]
LGTM will merge
